## Introduction
### Background
According to the Migration Policy Institute, more than 44,7 million immigrants lived in the United States in 2018. Then, according to Census.gov, 1.21 million people migrated to the US in 2018. Immigration has been a major source of population growth and cultural change throughout much of U.S. history. And one of the questions new arrivals are going to ask is where to settle down to start a new chapter in their life. We will try to answer this question on the example of Los Angeles.

### Problem
In this case, we are adopting machine learning tools to assist immigrants to find apartments with optimum conditions. The question we are currently solving is: how could we provide support to immigrants in finding suitable apartments in Los Angeles(further - LA)?
To solve this problem, we are going to cluster LA neighborhoods in order to recommend venues with the median renting price amount where immigrants can find suitable apartments. We will recommend neighborhoods according to essential facilities and services surrounding such venues i.e. schools, stores, restaurants, hospitals, and entertainments.


## Data
Data about Los Angeles(further - LA) neighborhoods and rent price paid in these neighborhoods was taken from USC Price Center for Social Innovation    (https://usc.data.socrata.com/Los-Angeles/Rent-Price-LA-/4a97-v5tx). The dataset comprises LA Neighborhoods, Median Rent Price Amount in LA Neighborhoods.

To discover and choose affordable and convenient locations across different neighborhoods according to the presence of essential facilities and services, we will access data through FourSquare API interface. By merging data frames on LA neighborhoods and the median rent price amount from USC price center and data on essential facilities and services surrounding these neighborhoods from FourSquare API interface, we will be able to recommend neighborhoods with affordable renting apartments.



## Methodology
#### The Methodology section is consist of  four following stages:

 1. Data Collection 
 2. Data Exploration and Understanding
 3. Data preparation and preprocessing 
 4. Modeling

## Importing Libraries
## Data Collection from USC Price Center for Social Innovation

In [1]:
import numpy as np

import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json

import requests
from pandas.io.json import json_normalize

import matplotlib.cm as cm
import matplotlib.colors as colors

from sklearn.cluster import KMeans

In [2]:
!pip install geopy
from geopy.geocoders import Nominatim

In [49]:
!conda install -c conda-forge folium=0.5.0 --yes
import folium

In [4]:
df = pd.read_csv('https://usc.data.socrata.com/api/views/4a97-v5tx/rows.csv')

## Data Exploration and Understanding

In [5]:
df.head()

Policy Area     Dataset           Variable  Year  Amount  \
0  Housing & Real Estate  Rent Price  Median Rent Price  2015  1115.0   
1  Housing & Real Estate  Rent Price  Median Rent Price  2015  1038.0   
2  Housing & Real Estate  Rent Price  Median Rent Price  2015  1031.0   
3  Housing & Real Estate  Rent Price  Median Rent Price  2015  1362.0   
4  Housing & Real Estate  Rent Price  Median Rent Price  2015   919.0   

                                               Tract  Tract Number  \
0  Census Tract 9200.23, Los Angeles County, Cali...        920023   
1  Census Tract 5344.04, Los Angeles County, Cali...        534404   
2  Census Tract 5336.01, Los Angeles County, Cali...        533601   
3  Census Tract 4053.02, Los Angeles County, Cali...        405302   
4  Census Tract 2042, Los Angeles County, California        204200   

    Neighborhood                 GEOID                    Location  \
0  Santa Clarita  1400000US06037920023  (34.4214975, -118.4905135)   
1         Cudahy  1400000US06037534404    (33.966232, -118.190553)   
2           Bell  1400000US06037533601   (33.9803325, -118.194495)   
3    West Covina  1400000US06037405302   (34.0767645, -117.946434)   
4  Boyle Heights  1400000US06037204200    (34.043615, -118.205084)   

                                        Row ID        Date  
0  Median_Rent_Price_2015_1400000US06037920023  01/01/2015  
1  Median_Rent_Price_2015_1400000US06037534404  01/01/2015  
2  Median_Rent_Price_2015_1400000US06037533601  01/01/2015  
3  Median_Rent_Price_2015_1400000US06037405302  01/01/2015  
4  Median_Rent_Price_2015_1400000US06037204200  01/01/2015

In [6]:
df.shape

(16390, 12)

In [7]:
df.dtypes

Policy Area      object
Dataset          object
Variable         object
Year              int64
Amount          float64
Tract            object
Tract Number      int64
Neighborhood     object
GEOID            object
Location         object
Row ID           object
Date             object
dtype: object

Dataset consist of 16390 rows and 12 columns. Now we can clean dataset and extract necessary columns 

##  Dataset preparation  and preprocessing
#### We prepare our dataset for the modeling process, by performing the following steps:
 1. Extract columns related to Neighborhoods, Median Rent Price and Location
 2. Rename the column name
 3. Select neighborhoods which median rent price fit the price range 
 4. Import venues around neighborhoods
 5. Join the data to find the coordinates of suitable locations 
 6. Plot recommended locations on LA map

#### Formating collected data

In [8]:
df = df[['Amount','Neighborhood', 'Location']]
df = df.rename(columns={'Amount':'Median Rent Price'})

In [9]:
df[['Latitude', 'Longitude']] = df['Location'].str.split(',', n=1, expand=True)

In [10]:
df = df.drop(['Location'], axis=1)

#### Preparing coordinates of neighboords

In [11]:
df['Latitude']= df['Latitude'].str.replace('(', '')
df['Longitude']= df['Longitude'].str.replace(')', '')
df.head()

Median Rent Price   Neighborhood    Latitude      Longitude
0             1115.0  Santa Clarita  34.4214975   -118.4905135
1             1038.0         Cudahy   33.966232    -118.190553
2             1031.0           Bell  33.9803325    -118.194495
3             1362.0    West Covina  34.0767645    -117.946434
4              919.0  Boyle Heights   34.043615    -118.205084

In [12]:
df[['Latitude', 'Longitude']] = df[['Latitude', 'Longitude']].astype('float64')
df.dtypes

Median Rent Price    float64
Neighborhood          object
Latitude             float64
Longitude            float64
dtype: object

#### Finding Median Rent Price among all Neighborhoods to narrow dataset

In [13]:
df.describe()

Median Rent Price      Latitude     Longitude
count       16055.000000  16390.000000  16390.000000
mean         1316.910308     34.065633   -118.251864
std           398.110637      0.179466      0.190607
min           100.000000     33.343563   -118.855281
25%          1031.000000     33.957001   -118.380005
50%          1213.000000     34.054677   -118.269599
75%          1520.000000     34.152711   -118.136926
max          3500.000000     34.792994   -117.708132

In [14]:
df1 = df[df['Median Rent Price'] >= 1315]
la_neighborhoods = df1[df1['Median Rent Price'] <= 1350]

#### Preparing final dataset

In [15]:
la_neighborhoods.head()

Median Rent Price Neighborhood   Latitude   Longitude
120             1348.0   Long Beach  33.822661 -118.176187
128             1330.0    Lancaster  34.711343 -118.157119
183             1333.0  Quartz Hill  34.638321 -118.221842
223             1336.0     Torrance  33.877806 -118.330794
243             1343.0   Sun Valley  34.211160 -118.371144

In [16]:
la_neighborhoods['Neighborhood'].value_counts()

Baldwin Park                             15
Pacoima                                  14
Long Beach                               14
Palmdale                                 12
Glendale                                 12
Alhambra                                 11
Downey                                   11
Monrovia                                 10
Pasadena                                 10
Monterey Park                             9
Azusa                                     8
Paramount                                 8
Burbank                                   8
Lakewood                                  8
Palms                                     8
Torrance                                  8
Bellflower                                8
West Covina                               8
Whittier                                  7
South Whittier                            7
Norwalk                                   7
Sun Valley                                6
Panorama City                   

In [17]:
la_neighborhoods.shape

(493, 4)

#### Getting the latitude and longitude values of Los Angeles

In [18]:
address = 'Los Angeles, CA'

geolocator = Nominatim(user_agent="la_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Los Angeles are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Los Angeles are 34.0536909, -118.2427666.


#### Creating a map of Los Angeles with neighborhoods superimposed on top

In [19]:
map_los_angeles = folium.Map(location=[latitude, longitude], zoom_start=11)


for lat, lng, label in zip(la_neighborhoods['Latitude'], la_neighborhoods['Longitude'], la_neighborhoods['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_los_angeles)  
    
map_los_angeles

#### Defining Foursquare Credentials and Version

In [20]:
CLIENT_ID = '41CSPZAYRM2LSTNHXVE0OVO5OJG1EGNFRNKURBFUZAK0YB1V' 
CLIENT_SECRET = 'ZFD1L1Z1VHJA4HUF2NISSU0PR2UMIA5KK3XC1WJRB5KKRBOZ' 
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 41CSPZAYRM2LSTNHXVE0OVO5OJG1EGNFRNKURBFUZAK0YB1V
CLIENT_SECRET:ZFD1L1Z1VHJA4HUF2NISSU0PR2UMIA5KK3XC1WJRB5KKRBOZ


# Modeling

#### Exploring Neighborhoods in Los Angeles

In [21]:
def getNearbyVenues(names, latitudes, longitudes, radius=200, LIMIT=50):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Creating Los Angeles Neighborhoods Venues dataframe (la_neighborhoods_venues)

In [23]:
la_neighborhoods_venues = getNearbyVenues(names=la_neighborhoods['Neighborhood'],
                                   latitudes=la_neighborhoods['Latitude'],
                                   longitudes=la_neighborhoods['Longitude']
                                  )

Long Beach
Lancaster
Quartz Hill
Torrance
Sun Valley
Pacoima
Santa Clarita
Baldwin Park
Artesia
Long Beach
Mid-Wilshire
Pacoima
Alhambra
Harbor City
Baldwin Park
Whittier
Fairfax
Industry
Rancho Park
South Whittier
San Pedro
Compton
Echo Park
Sun Valley
Downey
Norwalk
West Puente Valley
Elizabeth Lake
South San Jose Hills
Paramount
Pomona
El Monte
Sylmar
Santa Fe Springs
Sierra Madre
Palmdale
Inglewood
Azusa
Long Beach
Long Beach
Arcadia
Downey
San Pedro
Mar Vista
San Gabriel
South El Monte
Porter Ranch
Manchester Square
Bellflower
Winnetka
Norwalk
Pico Rivera
Alhambra
Paramount
Monrovia
South El Monte
El Monte
Montebello
Alhambra
West Covina
Alhambra
South San Jose Hills
Whittier
Montecito Heights
Palmdale
Alhambra
Palmdale
Monrovia
Pico Rivera
San Pasqual
Whittier Narrows
Claremont
Glendale
Hawaiian Gardens
Burbank
Highland Park
Burbank
Pomona
Downey
Bellflower
Downey
South Whittier
Lakewood
Sherman Oaks
Inglewood
Green Meadows
West Covina
Azusa
Pasadena
Alhambra
Silver Lake
Artesia


In [51]:
la_neighborhoods_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Quartz Hill              34.638321             -118.221842   
1   Sun Valley              34.211160             -118.371144   
2   Sun Valley              34.211160             -118.371144   
3      Artesia              33.859614             -118.079953   
4      Artesia              33.859614             -118.079953   

                  Venue  Venue Latitude  Venue Longitude   Venue Category  
0         The Brandwave       34.637017      -118.222812       Print Shop  
1  Active Key Insurance       34.210134      -118.370863           Office  
2      Vineland / Stagg       34.212018      -118.370521      Bus Station  
3              Sharetea       33.857890      -118.080290  Bubble Tea Shop  
4        Rainbow Donuts       33.858746      -118.081146           Bakery

#### Amount of venues were returned for each neighborhood

In [27]:
la_neighborhoods.groupby('Neighborhood').count()

Median Rent Price  Latitude  Longitude
Neighborhood                                                                 
Alhambra                                              11        11         11
Altadena                                               1         1          1
Arcadia                                                3         3          3
Arleta                                                 2         2          2
Artesia                                                6         6          6
Atwater Village                                        2         2          2
Avocado Heights                                        2         2          2
Azusa                                                  8         8          8
Baldwin Park                                          15        15         15
Bellflower                                             8         8          8
Beverly Grove                                          1         1          1
Boyle Heights                                          1         1          1
Brentwood                                              1         1          1
Broadway-Manchester                                    1         1          1
Burbank                                                8         8          8
Canoga Park                                            1         1          1
Carson                                                 2         2          2
Castaic Canyons                                        3         3          3
Charter Oak                                            5         5          5
Chatsworth                                             1         1          1
Citrus                                                 1         1          1
Claremont                                              2         2          2
Compton                                                1         1          1
Covina                                                 2         2          2
Culver City                                            3         3          3
Del Rey                                                2         2          2
Downey                                                11        11         11
Downtown                                               3         3          3
Duarte                                                 1         1          1
Eagle Rock                                             3         3          3
East San Gabriel                                       1         1          1
Echo Park                                              4         4          4
El Monte                                               6         6          6
El Segundo                                             2         2          2
Elizabeth Lake                                         2         2          2
Elysian Valley                                         1         1          1
Fairfax                                                1         1          1
Glendale                                              12        12         12
Glendora                                               3         3          3
Granada Hills                                          2         2          2
Green Meadows                                          1         1          1
Hacienda Heights                                       3         3          3
Harbor City                                            3         3          3
Hawaiian Gardens                                       1         1          1
Hawthorne                                              1         1          1
Highland Park                                          5         5          5
Hollywood                                              4         4          4
Hollywood Hills                                        2         2          2
Industry                                               4         4          4
Inglewood                                              3     

#### Analyzing Each Neighborhood

In [25]:
la_neighborhoods_onehot = pd.get_dummies(la_neighborhoods_venues[['Venue Category']], prefix="", prefix_sep="")

la_neighborhoods_onehot['Neighborhood'] = la_neighborhoods_venues['Neighborhood'] 

fixed_columns = [la_neighborhoods_onehot.columns[-1]] + list(la_neighborhoods_onehot.columns[:-1])
la_neighborhoods_onehot = la_neighborhoods_onehot[fixed_columns]

la_neighborhoods_onehot.head()

Zoo Exhibit  ATM  Accessories Store  American Restaurant  Antique Shop  \
0            0    0                  0                    0             0   
1            0    0                  0                    0             0   
2            0    0                  0                    0             0   
3            0    0                  0                    0             0   
4            0    0                  0                    0             0   

   Arcade  Argentinian Restaurant  Art Gallery  Art Museum  \
0       0                       0            0           0   
1       0                       0            0           0   
2       0                       0            0           0   
3       0                       0            0           0   
4       0                       0            0           0   

   Arts & Crafts Store  Arts & Entertainment  Asian Restaurant  \
0                    0                     0                 0   
1                    0                     0                 0   
2                    0                     0                 0   
3                    0                     0                 0   
4                    0                     0                 0   

   Athletics & Sports  Auto Workshop  Automotive Shop  BBQ Joint  Bagel Shop  \
0                   0              0                0          0           0   
1                   0              0                0          0           0   
2                   0              0                0          0           0   
3                   0              0                0          0           0   
4                   0              0                0          0           0   

   Bakery  Bank  Bar  Baseball Field  Beach  Bed & Breakfast  Beer Garden  \
0       0     0    0               0      0                0            0   
1       0     0    0               0      0                0            0   
2       0     0    0               0      0                0            0   
3       0     0    0               0      0                0            0   
4       1     0    0               0      0                0            0   

   Big Box Store  Boat or Ferry  Bookstore  Boutique  Bowling Alley  \
0              0              0          0         0              0   
1              0              0          0         0              0   
2              0              0          0         0              0   
3              0              0          0         0              0   
4              0              0          0         0              0   

   Boxing Gym  Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  \
0           0                     0               0        0                0   
1           0                     0               0        0                0   
2           0                     0               0        0                0   
3           0                     0               0        0                1   
4           0                     0               0        0                0   

   Buffet  Burger Joint  Bus Station  Bus Stop  Business Service  Café  \
0       0             0            0         0                 0     0   
1       0             0            0         0                 0     0   
2       0             0            1         0                 0     0   
3       0             0            0         0                 0     0   
4       0             0            0         0                 0     0   

   Cajun / Creole Restaurant  Campground  Candy Store  Cantonese Restaurant  \
0                          0           0            0                     0   
1                          0           0            0                     0   
2                          0           0            0                     0   
3                          0           0            0                     0   
4                          0           0            0                   

In [26]:
la_neighborhoods_onehot.shape

(1111, 224)

#### Frequency of occurrence of each category of venues in Neighborhoods

In [29]:
la_neighborhoods_grouped = la_neighborhoods_onehot.groupby('Neighborhood').mean().reset_index()
la_neighborhoods_grouped.head()

Neighborhood  Zoo Exhibit  ATM  Accessories Store  American Restaurant  \
0         Alhambra          0.0  0.0                0.0                  0.0   
1         Altadena          0.0  0.0                0.0                  0.0   
2          Arcadia          0.0  0.0                0.0                  0.0   
3          Artesia          0.0  0.0                0.0                  0.0   
4  Atwater Village          0.0  0.0                0.0                  0.0   

   Antique Shop  Arcade  Argentinian Restaurant  Art Gallery  Art Museum  \
0           0.0     0.0                     0.0          0.0         0.0   
1           0.0     0.0                     0.0          0.0         0.0   
2           0.0     0.0                     0.0          0.0         0.0   
3           0.0     0.0                     0.0          0.0         0.0   
4           0.0     0.0                     0.0          0.0         0.0   

   Arts & Crafts Store  Arts & Entertainment  Asian Restaurant  \
0                  0.0                   0.0               0.0   
1                  0.0                   0.0               0.0   
2                  0.0                   0.0               0.0   
3                  0.0                   0.0               0.0   
4                  0.0                   0.0               0.0   

   Athletics & Sports  Auto Workshop  Automotive Shop  BBQ Joint  Bagel Shop  \
0                 0.0            0.0              0.0        0.0         0.0   
1                 0.0            0.0              0.0        0.0         0.0   
2                 0.0            0.0              0.0        0.0         0.0   
3                 0.0            0.0              0.0        0.0         0.0   
4                 0.0            0.0              0.0        0.0         0.0   

     Bakery  Bank  Bar  Baseball Field  Beach  Bed & Breakfast  Beer Garden  \
0  0.000000   0.0  0.0             0.0    0.0              0.0          0.0   
1  0.000000   0.0  0.0             0.0    0.0              0.0          0.0   
2  0.000000   0.0  0.0             0.0    0.0              0.0          0.0   
3  0.074074   0.0  0.0             0.0    0.0              0.0          0.0   
4  0.000000   0.0  0.0             0.0    0.0              0.0          0.0   

   Big Box Store  Boat or Ferry  Bookstore  Boutique  Bowling Alley  \
0            0.0            0.0        0.0       0.0       0.076923   
1            0.0            0.0        0.0       0.0       0.000000   
2            0.0            0.0        0.0       0.0       0.000000   
3            0.0            0.0        0.0       0.0       0.000000   
4            0.0            0.0        0.0       0.0       0.000000   

   Boxing Gym  Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  \
0         0.0                   0.0        0.000000      0.0         0.000000   
1         0.0                   0.0        0.000000      0.0         0.000000   
2         0.0                   0.0        0.000000      0.0         0.000000   
3         0.0                   0.0        0.018519      0.0         0.148148   
4         0.0                   0.0        0.000000      0.0         0.000000   

   Buffet  Burger Joint  Bus Station  Bus Stop  Business Service  Café  \
0     0.0      0.153846          0.0       0.0               0.0   0.0   
1     0.0      0.000000          0.0       0.0               0.0   0.0   
2     0.0      0.000000          0.0       0.0               0.0   0.0   
3     0.0      0.000000          0.0       0.0               0.0   0.0   
4     0.0      0.000000          0.0       0.0               0.0   0.0   

   Cajun / Creole Restaurant  Campground  Candy Store  Cantonese Restaurant  \
0                        0.0         0.0          0.0                   0.0   
1                        0.0         0.0          0.0                   0.0   
2                        0.0         0.0          0.0                   0.0   
3                        0.0         0.0

In [32]:
la_neighborhoods_grouped.shape

(97, 224)

#### Finding Top 10 venues in each neighborhood

In [35]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [36]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))


neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = la_neighborhoods_grouped['Neighborhood']

for ind in np.arange(la_neighborhoods_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(la_neighborhoods_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0         Alhambra                 Hotel          Burger Joint   
1         Altadena          Intersection           Yoga Studio   
2          Arcadia                Garden      Sculpture Garden   
3          Artesia     Indian Restaurant       Bubble Tea Shop   
4  Atwater Village           Pizza Place          Liquor Store   

  3rd Most Common Venue   4th Most Common Venue 5th Most Common Venue  \
0              Pharmacy             Video Store                  Park   
1           Event Space  Furniture / Home Store     French Restaurant   
2      Cuban Restaurant  Furniture / Home Store     French Restaurant   
3            Playground      Chinese Restaurant        Ice Cream Shop   
4           Yoga Studio    Ethiopian Restaurant     French Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0              Nightclub        Shop & Service         Bowling Alley   
1       Football Stadium            Food Truck     Food & Drink Shop   
2       Football Stadium            Food Truck     Food & Drink Shop   
3  Vietnamese Restaurant        Sandwich Place          Dessert Shop   
4       Football Stadium            Food Truck     Food & Drink Shop   

  9th Most Common Venue 10th Most Common Venue  
0        Discount Store   Fast Food Restaurant  
1                  Food            Flea Market  
2                  Food            Flea Market  
3      Sushi Restaurant                 Bakery  
4                  Food            Flea Market

### Clustering Neighborhoods

#### Running k-means to cluster the neighborhood into 5 clusters

In [38]:
kclusters = 5

la_neighborhoods_grouped_clustering = la_neighborhoods_grouped.drop('Neighborhood', 1)

kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(la_neighborhoods_grouped_clustering)

kmeans.labels_[0:10]

array([3, 1, 3, 3, 3, 3, 3, 3, 3, 0], dtype=int32)

#### Creating a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood

In [39]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

la_neighborhoods_merged = la_neighborhoods

la_neighborhoods_merged = la_neighborhoods_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

la_neighborhoods_merged.head() 

Median Rent Price Neighborhood   Latitude   Longitude  Cluster Labels  \
120             1348.0   Long Beach  33.822661 -118.176187             3.0   
128             1330.0    Lancaster  34.711343 -118.157119             2.0   
183             1333.0  Quartz Hill  34.638321 -118.221842             3.0   
223             1336.0     Torrance  33.877806 -118.330794             3.0   
243             1343.0   Sun Valley  34.211160 -118.371144             3.0   

    1st Most Common Venue   2nd Most Common Venue       3rd Most Common Venue  \
120                  Park          Clothing Store          Italian Restaurant   
128                 Trail             Yoga Studio                 Event Space   
183            Print Shop               Jazz Club      Furniture / Home Store   
223                  Park          Clothing Store           Mobile Phone Shop   
243          Intersection  Furniture / Home Store  Construction & Landscaping   

      4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
120           Boat or Ferry           Gas Station  Gym / Fitness Center   
128  Furniture / Home Store     French Restaurant      Football Stadium   
183       French Restaurant      Football Stadium            Food Truck   
223     Japanese Restaurant        Ice Cream Shop    Seafood Restaurant   
243             Gas Station      Storage Facility          Burger Joint   

    7th Most Common Venue       8th Most Common Venue 9th Most Common Venue  \
120    Seafood Restaurant  Construction & Landscaping               Gay Bar   
128            Food Truck           Food & Drink Shop                  Food   
183     Food & Drink Shop                        Food           Flea Market   
223             Gift Shop               Grocery Store            Smoke Shop   
243         Grocery Store                      Office          Home Service   

    10th Most Common Venue  
120             Steakhouse  
128            Flea Market  
183      Fish & Chips Shop  
223        Udon Restaurant  
243            Bus Station

#### Creatina a map of resulting clusters

In [41]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

markers_colors = []
for lat, lon, poi, cluster in zip(la_neighborhoods_merged['Latitude'], la_neighborhoods_merged['Longitude'], la_neighborhoods_merged['Neighborhood'], la_neighborhoods_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color='#3186cc',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Clusters Examination

### Cluster 1

In [44]:
la_neighborhoods_merged.loc[la_neighborhoods_merged['Cluster Labels'] == 0, la_neighborhoods_merged.columns[[1] + list(range(5, la_neighborhoods_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
3775         Larchmont             Speakeasy                   Gym   
6926         Brentwood             Speakeasy      Cuban Restaurant   
13497  Hollywood Hills             BBQ Joint           Yoga Studio   
14106  Hollywood Hills             BBQ Joint           Yoga Studio   

      3rd Most Common Venue   4th Most Common Venue 5th Most Common Venue  \
3775              BBQ Joint             Gas Station                Garden   
6926                 Garden  Furniture / Home Store     French Restaurant   
13497           Gas Station  Furniture / Home Store     French Restaurant   
14106           Gas Station  Furniture / Home Store     French Restaurant   

        6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
3775   Furniture / Home Store     French Restaurant      Football Stadium   
6926         Football Stadium            Food Truck     Food & Drink Shop   
13497        Football Stadium            Food Truck     Food & Drink Shop   
14106        Football Stadium            Food Truck     Food & Drink Shop   

      9th Most Common Venue 10th Most Common Venue  
3775             Food Truck      Food & Drink Shop  
6926                   Food            Flea Market  
13497                  Food            Flea Market  
14106                  Food            Flea Market

### Cluster 2

In [45]:
la_neighborhoods_merged.loc[la_neighborhoods_merged['Cluster Labels'] == 1, la_neighborhoods_merged.columns[[1] + list(range(5, la_neighborhoods_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
13038     Altadena          Intersection           Yoga Studio   
16022   Wilmington        Baseball Field          Intersection   
16183    Hawthorne          Intersection        Baseball Field   

      3rd Most Common Venue   4th Most Common Venue   5th Most Common Venue  \
13038           Event Space  Furniture / Home Store       French Restaurant   
16022           Yoga Studio             Event Space  Furniture / Home Store   
16183                  Park             Yoga Studio                    Farm   

        6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
13038        Football Stadium            Food Truck     Food & Drink Shop   
16022       French Restaurant      Football Stadium            Food Truck   
16183  Furniture / Home Store     French Restaurant      Football Stadium   

      9th Most Common Venue 10th Most Common Venue  
13038                  Food            Flea Market  
16022     Food & Drink Shop                   Food  
16183            Food Truck      Food & Drink Shop

### Cluster 3

In [46]:
la_neighborhoods_merged.loc[la_neighborhoods_merged['Cluster Labels'] == 2, la_neighborhoods_merged.columns[[1] + list(range(5, la_neighborhoods_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
128         Lancaster                 Trail           Yoga Studio   
981         Echo Park                 Trail           Yoga Studio   
6590        Echo Park                 Trail           Yoga Studio   
8243   Ladera Heights                 Trail           Yoga Studio   
8729        Lancaster                 Trail           Yoga Studio   
9941        Echo Park                 Trail           Yoga Studio   
11388       Lancaster                 Trail           Yoga Studio   
12753       Lancaster                 Trail           Yoga Studio   
12938       Echo Park                 Trail           Yoga Studio   
13347       Lancaster                 Trail           Yoga Studio   
13707  Ladera Heights                 Trail           Yoga Studio   
15897  Ladera Heights                 Trail           Yoga Studio   

      3rd Most Common Venue   4th Most Common Venue 5th Most Common Venue  \
128             Event Space  Furniture / Home Store     French Restaurant   
981             Event Space  Furniture / Home Store     French Restaurant   
6590            Event Space  Furniture / Home Store     French Restaurant   
8243            Event Space  Furniture / Home Store     French Restaurant   
8729            Event Space  Furniture / Home Store     French Restaurant   
9941            Event Space  Furniture / Home Store     French Restaurant   
11388           Event Space  Furniture / Home Store     French Restaurant   
12753           Event Space  Furniture / Home Store     French Restaurant   
12938           Event Space  Furniture / Home Store     French Restaurant   
13347           Event Space  Furniture / Home Store     French Restaurant   
13707           Event Space  Furniture / Home Store     French Restaurant   
15897           Event Space  Furniture / Home Store     French Restaurant   

      6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
128        Football Stadium            Food Truck     Food & Drink Shop   
981        Football Stadium            Food Truck     Food & Drink Shop   
6590       Football Stadium            Food Truck     Food & Drink Shop   
8243       Football Stadium            Food Truck     Food & Drink Shop   
8729       Football Stadium            Food Truck     Food & Drink Shop   
9941       Football Stadium            Food Truck     Food & Drink Shop   
11388      Football Stadium            Food Truck     Food & Drink Shop   
12753      Football Stadium            Food Truck     Food & Drink Shop   
12938      Football Stadium            Food Truck     Food & Drink Shop   
13347      Football Stadium            Food Truck     Food & Drink Shop   
13707      Football Stadium            Food Truck     Food & Drink Shop   
15897      Football Stadium            Food Truck     Food & Drink Shop   

      9th Most Common Venue 10th Most Common Venue  
128                    Food            Flea Market  
981                    Food            Flea Market  
6590                   Food            Flea Market  
8243                   Food            Flea Market  
8729                   Food            Flea Market  
9941                   Food            Flea Market  
11388                  Food            Flea Market  
12753                  Food            Flea Market  
12938                  Food            Flea Market  
13347                  Food            Flea Market  
13707                  Food            Flea Market  
15897                  Food            Flea Market

### Cluster 4

In [47]:
la_neighborhoods_merged.loc[la_neighborhoods_merged['Cluster Labels'] == 3, la_neighborhoods_merged.columns[[1] + list(range(5, la_neighborhoods_merged.shape[1]))]]

Neighborhood      1st Most Common Venue  \
120              Long Beach                       Park   
183             Quartz Hill                 Print Shop   
223                Torrance                       Park   
243              Sun Valley               Intersection   
254                 Pacoima                Supermarket   
265            Baldwin Park  Latin American Restaurant   
296                 Artesia          Indian Restaurant   
321              Long Beach                       Park   
329            Mid-Wilshire                        Gym   
371                 Pacoima                Supermarket   
455                Alhambra                      Hotel   
548            Baldwin Park  Latin American Restaurant   
554                Whittier                       Pool   
567                 Fairfax          Convenience Store   
597                Industry          Electronics Store   
607             Rancho Park           Pedestrian Plaza   
753          South Whittier                 Food Truck   
839               San Pedro                       Park   
1013             Sun Valley               Intersection   
1074                 Downey                    Brewery   
1087                Norwalk               Intersection   
1185   South San Jose Hills         Golf Driving Range   
1194              Paramount                       Park   
1241                 Pomona                       Park   
1255               El Monte              Deli / Bodega   
1277       Santa Fe Springs          Electronics Store   
1454             Long Beach                       Park   
1461             Long Beach                       Park   
1571                Arcadia                     Garden   
1585                 Downey                    Brewery   
1595              San Pedro                       Park   
1602              Mar Vista                        Spa   
1696            San Gabriel                       Park   
1841      Manchester Square               Burger Joint   
1893             Bellflower                Zoo Exhibit   
1914               Winnetka              Women's Store   
1941                Norwalk               Intersection   
1990            Pico Rivera                Coffee Shop   
2019               Alhambra                      Hotel   
2026              Paramount                       Park   
2051               Monrovia                     Garden   
2134               El Monte              Deli / Bodega   
2193             Montebello        American Restaurant   
2222               Alhambra                      Hotel   
2237            West Covina                       Park   
2263               Alhambra                      Hotel   
2274   South San Jose Hills         Golf Driving Range   
2288               Whittier                       Pool   
2320      Montecito Heights            Thai Restaurant   
2387               Alhambra                      Hotel   
2414               Monrovia                     Garden   
2551            Pico Rivera                Coffee Shop   
2553            San Pasqual        American Restaurant   
2647              Claremont      Performing Arts Venue   
2666               Glendale         Chinese Restaurant   
2700       Hawaiian Gardens              Auto Workshop   
2728                Burbank              Deli / Bodega   
2761          Highland Park                Art Gallery   
2782                Burbank              Deli / Bodega   
2806                 Pomona                       Park   
2883                 Downey                    Brewery   
2904             Bellflower                Zoo Exhibit   
2949                 Downey                    Brewery   
3026         South Whittier                 Food Truck   
3036               Lakewood              Moving Target   
3100           Sherman Oaks              Moving Target   
3261          Green Meadows                     Lawyer   
3284            West Covina                       Park   
3305         

### Cluster 5

In [48]:
la_neighborhoods_merged.loc[la_neighborhoods_merged['Cluster Labels'] == 4, la_neighborhoods_merged.columns[[1] + list(range(5, la_neighborhoods_merged.shape[1]))]]

Neighborhood       1st Most Common Venue 2nd Most Common Venue  \
863          Compton  Construction & Landscaping           Yoga Studio   
5383   Granada Hills  Construction & Landscaping            Food Truck   
7703   Granada Hills  Construction & Landscaping            Food Truck   
15935   San Fernando  Construction & Landscaping           Yoga Studio   

      3rd Most Common Venue   4th Most Common Venue   5th Most Common Venue  \
863             Event Space  Furniture / Home Store       French Restaurant   
5383            Yoga Studio             Event Space  Furniture / Home Store   
7703            Yoga Studio             Event Space  Furniture / Home Store   
15935           Event Space  Furniture / Home Store       French Restaurant   

      6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
863        Football Stadium            Food Truck     Food & Drink Shop   
5383      French Restaurant      Football Stadium     Food & Drink Shop   
7703      French Restaurant      Football Stadium     Food & Drink Shop   
15935      Football Stadium            Food Truck     Food & Drink Shop   

      9th Most Common Venue 10th Most Common Venue  
863                    Food            Flea Market  
5383                   Food            Flea Market  
7703                   Food            Flea Market  
15935                  Food            Flea Market

# Result  and Discussion section

We examined this dataset using two main perspectives.
The first is finding neighborhoods that average rent price fit Median Rent Price among all Neighborhoods.
The second is clustering found neighborhoods using data about venues located around them.
 
As a result, we can see that the Neighborhoods from Cluster 4 can be considered more suitable for newcomers to rent apartments, because they contain a big variety of different venues in close range which can satisfy their needs. Other Clusters are not so diversified and can be suitable for immigrants who have enough experience of living in Los Angeles and familiar with the city. 


# Conclusion

According to the Migration Policy Institute, more than 14% of the population of the United States were considered immigrants in 2018. From which we can assume that immigrants have become a major part of the US society and one of the drivers of cultural changes. In this context of a continuous influx of migrants it is urgent to adopt machine learning tools in order to assist newcomers to find a suitable place to settle down. As a result, the question we were trying to solve was: how could we provide support to immigrants in finding suitable apartments in example of Los Angeles? 

To solve this problem, we clustered LA neighborhoods in order to recommend neighborhoods with the median renting price amount where immigrants can find suitable apartments. We recommended neighborhoods according to essential facilities and services surrounding such venues i.e. schools, stores, restaurants, hospitals, and entertainments.

We gathered LA Neighborhoods and the Median Rent Price data from USC Price Center for Social Innovation (https://usc.data.socrata.com/Los-Angeles/Rent-Price-LA-/4a97-v5tx). Moreover, to explore and target recommended locations across different venues according to the presence of essential facilities and services we accessed data through the FourSquare API interface and arranged them as a data frame for visualization. By merging LA Neighborhoods and the Median Rent Price data and data on essential facilities and services surrounding such properties from the FourSquare API interface, we were able to recommend suitable neighborhoods to rent apartments.

The Methodology section comprised four stages: 1. Data Collection; 2. Data Exploration and Understanding; 3. Data preparation and preprocessing; 4. Modeling. In the modeling section, we used the k-means clustering technique as it is highly flexible to account for changes in the median rent price amount in Los Angeles Neighborhoods.

Finally, we drew the conclusion suggesting suitable neighborhoods to rent apartments. We used two main perspectives to reach our results. First, we examined them according to median price amount paid around all neighborhoods. Second, we analyzed our results according to the five clusters we produced and found that Cluster 4 might be considered as the most suitable.